### ISSUE TITLE
migrate BicyclePumpNode to scenery-phet

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/states-of-matter#217]
The Gas Properties design document description of the bicycle pump UI component (see screenshot below) points to "similar in States of Matter".   I see `STATES_OF_MATTER/BicyclePumpNode`, but it's coupled to that sim.   So we'll need to allocate a chunk of time to generalize it and move it to scenery-phet, so that I can use it in Gas Properties.

Assigning to me, @jbphet and @ariel-phet to discuss.

![screenshot_702](https://user-images.githubusercontent.com/3046552/42838582-7346d05c-89be-11e8-9fe4-5c23594c2c39.png)

**Comment by jbphet** [comment]
Porting this to common code would probably be a decent introduction for @chrisklus.  We could have a short discussion with @pixelzoom, @chrisklus, and me to decide what generalization is needed and then I could support him during the work.  The original port was done by Actual Concepts, then refined by me, and there are some things that could likely be simplified.

**Comment by jbphet** [comment]
I have a pretty clear list of what needs to be done on SOM to get it published, so we can discuss at any time.  Probably the biggest effort is some generalization of the color scheme handling.

**Comment by ariel-phet** [comment]
Since we agree this would be a nice intro to common code for @chrisklus we will have @jbphet on this with him when @jbphet dives back into trying to get SOM republished

**Comment by pixelzoom** [comment]
@ariel-phet I need this sometime in the next couple of weeks.

**Comment by pixelzoom** [comment]
Heads up! The pump in Gas Properties deviates from the pump in SOM.   I don't fully understand the implications of this yet, but here's the comment thread from the Gas Properties design doc: 

> @pixelzoom: How many particles are added? Are more particles added based on how far the handle moves? If so, what is the range of how many particles are added?
> 
> @arouinfar: In Java, the number depends on how far/fast the handle moves. I think we'll want similar behavior in HTML5. Perhaps an appropriate max number of particles/pump is 10-20% of the max number of particles. With the current ranges, that would be a maximum of 100-200 particles/pump. We can confirm the exact number in design meeting.
> 
> @arouinfar: As a heads up, this is a deviation from the way bicyclePumpNode behaves in SOM, which always adds a set number of particles. When migrating the pump to common code, there will need to be options for either adding a set number or tying the number to the dynamics of the handle.

**Comment by arouinfar** [comment]
Beat me to it, thanks @pixelzoom. 

I think we'll need to have some design discussion about how the dynamics of the handle should affect the number of particles added, so I'll open an issue in gas-properties for that.

**Comment by pixelzoom** [comment]
@chrisklus Be sure to see related discussion over in https://github.com/phetsims/gas-properties/issues/33.

**Comment by pixelzoom** [comment]
I don't know how SOM implemented the pump.  But in Gas Properties, I'm planning on having 1 pump for each type of particle ('heavy' and 'light'), then switching between them based on which type of particle is selected via radio buttons.  Mockup from the design doc:

![screenshot_859](https://user-images.githubusercontent.com/3046552/48443092-4d4a1980-e74d-11e8-9140-f0b251e0cde5.png)

**Comment by chrisklus** [comment]
@pixelzoom do you know the approximate time when it would be nice for this to be done? I think it's going to take me a long time so I'm trying to plan out how much I should allocate per week (or day, if needed very soon).

**Comment by pixelzoom** [comment]
From the design document:

> **Milestones**
> January 2019:  Design progress will be included in grant report. If there is a workable dev, that might be included, but not required.
> January 2020: Grant report is due, so sim must be published

That doesn't give me any meaningful milestones for when we might need to show this to anyone outside of PhET.  So I'm going to ask @arouinfar: When do we need to show this sim to an audience outside of PhET?   I'd like to have a working BicyclePumpNode around 2 weeks before that date.

**Comment by pixelzoom** [comment]
Thanks @arouinfar.

@chrisklus Let's shoot for completing this work by the end of March 2019.  Does that seem reasonable?

**Comment by chrisklus** [comment]
Thanks @pixelzoom and @arouinfar, yes @pixelzoom that sounds great. @SaurabhTotey and I will continue to chip away, and I'll add in more time as needed.

I was originally concerned because of https://github.com/phetsims/states-of-matter/issues/217#issuecomment-438118562.

**Comment by pixelzoom** [comment]
Mentioned at last week's dev meeting...  We should discuss how the ends of the hose are attached -- to the pump, and to whatever is on the other end.   I'm not sure if this should be the responsibility of `BicyclePumpNode` or the client.  It would be nice to have a `HoseNode` (similar to `WireNode`) that makes this easy.  And it should be possible to attach the hose to either side of the pump.

**Comment by pixelzoom** [comment]
My WAG for when I'd need BicyclePumpNode was "sometime in April" and that's turning out to be right on the money.

In https://github.com/phetsims/states-of-matter/issues/217#issuecomment-442282755, @arouinfar indicated that it would be nice to show off this sim at NSTA (April 11-14) or ISTE (June 23-26), if the first 2 screens are ready.  Given where it's at now, that's still a possibility. 

But as I said in https://github.com/phetsims/states-of-matter/issues/217#issuecomment-442277409:
>  I'd like to have a working BicyclePumpNode around 2 weeks before that date.

That means that this issue needs to be completed by April 4.  @chrisklus is that possible?

**Comment by pixelzoom** [comment]
In In https://github.com/phetsims/states-of-matter/issues/217#issuecomment-442283609, my WAG for when this would need to be completed was:
> Let's shoot for completing this work by the end of March 2019. Does that seem reasonable? 

In https://github.com/phetsims/states-of-matter/issues/217#issuecomment-442282755, @arouinfar indicated that it would be nice to show off this sim at NSTA (April 11-14) or ISTE (June 23-26), if the first 2 screens are ready.  Given where it's at now, that's still a possibility. 

But as I said in https://github.com/phetsims/states-of-matter/issues/217#issuecomment-442277409:
>  I'd like to have a working BicyclePumpNode around 2 weeks before that date.

That means that this issue needs to be completed by April 4.  @chrisklus is that possible?

**Comment by chrisklus** [comment]
>That means that this issue needs to be completed by April 4. @chrisklus is that possible?

@pixelzoom yep! Good timing.

**Comment by kathy-phet** [comment]
:) ... Happy to hear its still possible to show off the first 2 screens by these upcoming conferences!

**Comment by pixelzoom** [comment]
@chrisklus FYI, See discussion about color-related parameters/options in https://github.com/phetsims/scenery/issues/962#issuecomment-480034075.

**Comment by chrisklus** [comment]
@pixelzoom I've been having a hard time fixing the buggy behavior of the "number of particles added per pump action" situation. Could you take a look at this patch?

<details>

```js
Index: js/common/view/BicyclePumpNode.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/common/view/BicyclePumpNode.js	(revision 781afefd5a64aef8419760510176fd71becde83d)
+++ js/common/view/BicyclePumpNode.js	(date 1554961489000)
@@ -14,13 +14,13 @@
   // modules
   const BooleanProperty = require( 'AXON/BooleanProperty' );
   const Color = require( 'SCENERY/util/Color' );
+  const DragListener = require( 'SCENERY/listeners/DragListener' );
   const LinearGradient = require( 'SCENERY/util/LinearGradient' );
   const Node = require( 'SCENERY/nodes/Node' );
   const Path = require( 'SCENERY/nodes/Path' );
   const Rectangle = require( 'SCENERY/nodes/Rectangle' );
   const SegmentedBarGraphNode = require( 'STATES_OF_MATTER/common/view/SegmentedBarGraphNode' );
   const Shape = require( 'KITE/Shape' );
-  const SimpleDragHandler = require( 'SCENERY/input/SimpleDragHandler' );
   const statesOfMatter = require( 'STATES_OF_MATTER/statesOfMatter' );
   const Util = require( 'DOT/Util' );
   const Vector2 = require( 'DOT/Vector2' );
@@ -90,8 +90,6 @@
       // @private - used to track where the current position is on the handle when drawing its gradient
       this.handleGradientPosition = 0;
 
-      let currentPumpingDistance = 0;
-
       // create the base of the pump
       const baseWidth = width * PUMP_BASE_WIDTH_PROPORTION;
       const baseHeight = height * PUMP_BASE_HEIGHT_PROPORTION;
@@ -183,44 +181,10 @@
       const maxHandleYOffset = pumpHandleNode.centerY;
       const minHandleYOffset = maxHandleYOffset + ( -PUMP_SHAFT_HEIGHT_PROPORTION * height / 2 );
 
-      // How far the pump shaft needs to travel before the pump releases a particle. -1 is added to account for minor drag
-      // listener and floating-point errors.
-      const pumpingDistanceRequiredToAddParticle = ( -minHandleYOffset + maxHandleYOffset ) /
-                                                   options.numberOfParticlesPerPumpAction - 1;
-
-      pumpHandleNode.addInputListener( new SimpleDragHandler( {
-
-        drag: event => {
-
-          // the position at the start of the drag event
-          const handleStartYPos = pumpHandleNode.centerY;
-
-          // update the handle and shaft position based on the user's pointer position
-          const dragPositionY = pumpHandleNode.globalToParentPoint( event.pointer.point ).y;
-          pumpHandleNode.centerY = Util.clamp( dragPositionY, minHandleYOffset, maxHandleYOffset );
-          pumpShaftNode.top = pumpHandleNode.bottom;
-
-          const travelDistance = handleStartYPos - pumpHandleNode.centerY;
-          if ( travelDistance < 0 ) {
-
-            // This motion is in the downward direction, so add its distance to the pumping distance.
-            currentPumpingDistance += Math.abs( travelDistance );
-            while ( currentPumpingDistance >= pumpingDistanceRequiredToAddParticle ) {
-
-              // Enough distance has been traveled to inject a new particle.
-              if ( rangeProperty.value.max - numberProperty.value > 0 && options.enabledProperty.get() ) {
-                numberProperty.value++;
-              }
-              currentPumpingDistance -= pumpingDistanceRequiredToAddParticle;
-            }
-          }
-          else if ( travelDistance > 0 ) {
-
-            // This motion is in the upward direction, so reset any accumulated pumping distance.
-            currentPumpingDistance = 0;
-          }
-        }
-      } ) );
+      pumpHandleNode.addInputListener(
+        new HandleNodeDragListener( numberProperty, rangeProperty, options.enabledProperty,
+          minHandleYOffset, maxHandleYOffset, pumpHandleNode, pumpShaftNode, options.numberOfParticlesPerPumpAction )
+      );
 
       // create the node that will be used to indicate the remaining capacity
       const remainingCapacityIndicator = new SegmentedBarGraphNode(
@@ -535,5 +499,74 @@
     }
   }
 
-  return statesOfMatter.register( 'BicyclePumpNode', BicyclePumpNode );
+  statesOfMatter.register( 'BicyclePumpNode', BicyclePumpNode );
+
+  /**
+   * Drag listener for the pump's handle.
+   */
+  class HandleNodeDragListener extends DragListener {
+
+    /**
+     *
+     * @param numberProperty
+     * @param rangeProperty
+     * @param enabledProperty
+     * @param minHandleYOffset
+     * @param maxHandleYOffset
+     * @param pumpHandleNode
+     * @param pumpShaftNode
+     * @param numberOfParticlesPerPumpAction
+     */
+    constructor( numberProperty,
+                 rangeProperty,
+                 enabledProperty,
+                 minHandleYOffset,
+                 maxHandleYOffset,
+                 pumpHandleNode,
+                 pumpShaftNode,
+                 numberOfParticlesPerPumpAction
+    ) {
+
+      let handlePosition = null;
+      let lastHandlePosition = 0;
+      let pumpingDistanceAccumulation = 0;
+
+      // How far the pump shaft needs to travel before the pump releases a particle. -1 is added to account for minor drag
+      // listener and floating-point errors.
+      const pumpingDistanceRequiredToAddParticle = ( -minHandleYOffset + maxHandleYOffset ) /
+                                                   numberOfParticlesPerPumpAction - 1;
+
+      super( {
+        drag: ( event, listener ) => {
+
+          // update the handle and shaft position based on the user's pointer position
+          const dragPositionY = pumpHandleNode.globalToParentPoint( event.pointer.point ).y;
+          handlePosition = Util.clamp( dragPositionY, minHandleYOffset, maxHandleYOffset );
+          pumpHandleNode.centerY = handlePosition;
+          pumpShaftNode.top = pumpHandleNode.bottom;
+
+          const travelDistance = handlePosition - lastHandlePosition;
+          if ( travelDistance > 0 ) {
+
+            // This motion is in the downward direction, so add its distance to the pumping distance.
+            pumpingDistanceAccumulation += travelDistance;
+            while ( pumpingDistanceAccumulation >= pumpingDistanceRequiredToAddParticle ) {
+
+              // inject a particle
+              if ( rangeProperty.value.max - numberProperty.value > 0 && enabledProperty.get() ) {
+                numberProperty.value++;
+              }
+              pumpingDistanceAccumulation -= pumpingDistanceRequiredToAddParticle;
+            }
+          }
+
+          lastHandlePosition = handlePosition;
+        }
+      } );
+    }
+  }
+
+  statesOfMatter.register( 'BicyclePumpNode.HandleNodeDragListener', HandleNodeDragListener );
+
+  return BicyclePumpNode;
 } );
```

</details>

My intent is for it to handle the varying sizes of dx that come from different pumping speeds. And it does so pretty well for a limited range of speeds, but not well at all for fast speeds. Observe the behavior when `numberOfParticlesPerPumpAction` is 8 - at the fastest speed I can test, only 3 particles are added. Am I implementing this pattern incorrectly?

Maybe I'm going about this the wrong way and should instead calculate fixed points in the drag range where new particles should be added? Or something of that nature that isn't as reliant on getting drag events?

**Comment by pixelzoom** [comment]
@chrisklus I'll have a look, but it might not be until next week.

**Comment by pixelzoom** [comment]
@chrisklus and I collaborated on `HandleNodeDragListener`.  It appears to be operating correctly, and there problem with number of particles added is in SOM.   

Assign back to me when ready for next steps.

**Comment by pixelzoom** [comment]
For the purposes of connecting the hose to the container in Gas Properties, here's what that looks like in the design doc:

<img width="782" alt="screenshot_1159" src="https://user-images.githubusercontent.com/3046552/56767117-2ebe6480-6768-11e9-8c7d-a78cc75ba66d.png">

**Comment by chrisklus** [comment]
@pixelzoom ready for review.

In https://github.com/phetsims/gas-properties/issues/59, I integrated BicyclePumpNode (BPN) into Gas Properties, so check out that branch if you'd like. I moved the injection point down to where it is in the mockup.

In https://github.com/phetsims/scenery-phet/commit/cef194e3f713063ccd1c8dcdeef8e6bec6e325b2, I added a small constant to fix the calculation for `numberOfParticlesPerPumpAction`, because without it, SOM was pumping out 2 and Gas Props was pumping out 19 when set to 3 and 20, respectively. I tested various numbers between 1 and 30 and this constant appears to fix those cases.

I also added `reset()` that puts the handle back in the starting position, which SOM's implementation did not have.

There is no dispose yet, since I would think that BPNs will live for the life of the sim. Let me know if you disagree.

**Comment by pixelzoom** [comment]
I've integrated BicyclePumpNode into master.   It was a little time-consuming because the changes in bicycle-pump branch contained lots of formatting changes, so it was a little difficult to identify the relevant changes.  But I got it sorted out.  

The first thing I noticed is the excessive touchArea for the pump handle.  So I added options for pointer areas in the above commit:

```js
        // pointer areas
        handleTouchAreaXDilation: 15,
        handleTouchAreaYDilation: 15,
        handleMouseAreaXDilation: 0,
        handleMouseAreaYDilation: 0
```

I set them to the old values in SOM/PhaseChangesScreenView:

```js
146        handleTouchAreaXDilation: 100,
147        handleTouchAreaYDilation: 100
```

A couple of things that we need to work on:

(1) `setHoseAttachmentPosition` looks like it should be doable through options. Having to make a separate call is inconvenient.

(2) The relationship between `options.hoseAttachmentOffset` and `setHoseAttachmentPosition` feels backwards.  I should be able to set the location of the pump (e.g. bottom center of the base), then specify the offset of the hose.

(3) We're going to need some control over the shape of the hose.  The current look is a little too curvy for Gas Properties.   If we can't come up with a general "look" that is acceptable, then we may need to add option(s) for control points.

**Comment by pixelzoom** [comment]
(4) Width and height don't give me a bigger pump, they just scale the pump.   So if I want a pump that is a little taller or short (by changing `options.height`) then the _entire pump_ scales down, in all dimensions.  So there's no way for me to have a longer handle, or a taller cylinder.   This may be OK for now, but I don't think it's going to cut it longterm.  I'd like to hear what designers (and especially @ariel-phet) have to say.

**Comment by pixelzoom** [comment]
(5) The origin is in an inconvenient location, see red dot in the screenshot below.  For coordinating positioning of the pump and the hose, `centerBottom` would be a better place for the origin.

<img width="112" alt="screenshot_1167" src="https://user-images.githubusercontent.com/3046552/57004390-e8b03900-6b8b-11e9-90fd-fbd5c866481e.png">

**Comment by pixelzoom** [comment]
I the above commits, I took care of (1) and (2).  I eliminated `setHoseAttachmentPosition`.  You position a pump using the typical scenery translation options.  Then you provide `options.hoseAttachmentOffset` to indicate where the far end of the hose is, relative to the pump position.  See example in SOM and Gas Properties.  Here's an abbreviated version of GasPropertiesScreenView:

```js
// Bicycle pump is centered above the radio buttons.
const bicyclePumpLocation = new Vector2( particleTypeRadioButtonGroup.centerX, particleTypeRadioButtonGroup.top - 30 );

// Bicycle pump hose attaches to the container.
const hoseAttachmentPoint = new Vector2( containerNode.right,
  model.modelViewTransform.modelToViewY( model.container.hoseLocation.y ) );

const heavyBicyclePumpNode = new GasPropertiesBicyclePumpNode( model.numberOfHeavyParticlesProperty, {
  translation: bicyclePumpLocation,
  hoseAttachmentOffset: hoseAttachmentPoint.minus( bicyclePumpLocation ),
  ..
} );
```

Does this look acceptable?

**Comment by pixelzoom** [comment]
Re (3) shape of the hose... 

If I change this line in BicyclePumpNode

```js
173    .cubicCurveTo( 1.5 * ...
```

to this:

```js
173    .cubicCurveTo( 0.75 * ...
```

.. then things look much better in Gas Properties (and arguably in SOM).  

In Gas Properties, here's before (left, 1.5) and after (right, 0.75):

<img width="182" alt="screenshot_1168" src="https://user-images.githubusercontent.com/3046552/57005170-cc63ca80-6b92-11e9-8f45-d863bbe0bf1f.png">  <img width="147" alt="screenshot_1170" src="https://user-images.githubusercontent.com/3046552/57005172-d1287e80-6b92-11e9-896e-98619bfa6a15.png">

In SOM, here's before (left, 1.5) and after (right, 0.75):

<img width="274" alt="screenshot_1172" src="https://user-images.githubusercontent.com/3046552/57005219-42683180-6b93-11e9-984f-ee46cf1fbd15.png">  <img width="284" alt="screenshot_1171" src="https://user-images.githubusercontent.com/3046552/57005220-45fbb880-6b93-11e9-9cd6-69122fd2b0f0.png">



We can either change this multiplier to 0.75, or add an option (name suggestions?)

**Comment by chrisklus** [comment]
>You position a pump using the typical scenery translation options. Then you provide options.hoseAttachmentOffset to indicate where the far end of the hose is, relative to the pump position. Does this look acceptable?

Yep, that looks great, thanks for making the change.

>Re (3) shape of the hose... ...add an option (name suggestions?)

Indeed, I was thinking we might need to expose the control points with options. It seems like the desired values will vary depending on the distance that the hose extends away from the pump. But, given your demonstration in https://github.com/phetsims/states-of-matter/issues/217#issuecomment-488205854, it seems like we gain a fair amount of flexibility with changing just that one multiplier (instead of messing with 2-4 control point values). What do you think about something like (best name I can think of so far):

```js
hoseCurvynessMultiplier: 1 // higher number = curvy hose, lower number = straighter hose
```

>(4) Changing options.height doesn't give me a bigger pump, it just scales the pump. So if I want a pump that is a little taller or short (by changing options.height) then the entire pump scales down, in all dimensions. So there's no way for me to have a longer handle, or a taller cylinder. This may be OK for now, but I don't think it's going to cut it longterm. I'd like to hear what designers (and especially @ariel-phet) have to say.

Right, yeah this was the way the pump was originally coded - I'm not crazy about it either. I could remove all of those multiplier constants at the beginning of the file, replace all of the drawings with absolute numbers (instead of calculated percentages), and then expose a few dimensions via options?

e.g.
```js
handleWidth: 200,
handleHeight: 30,
bodyHeight: 450
```

>(5) The origin is in an inconvenient location, see red dot in the screenshot below. For coordinating positioning of the pump and the hose, centerBottom would be a better place for the origin. (EDIT: When running with ?dev, I've added a red dot at the origin.)

Whoops, looks like I was a little off when redrawing the components to position the whole node from the center bottom. I'll correct this to be the true center bottom.

**Comment by pixelzoom** [comment]
> hoseCurvynessMultiplier: 1

That sounds fine.

> I could remove all of those multiplier constants at the beginning of the file, replace all of the drawings with absolute numbers (instead of calculated percentages), and then expose a few dimensions via options?

Check with @ariel-phet or some of the designers.  If the designers are happy with size/proportions of the pump in Gas Properties, I'm happy.    If we were going to do this, I would recommend options:
`handleWidth, handleHeight, bodyWidth, bodyHeight`, then scale the bottom piece based on `bodyWidth`. 

> Whoops, looks like I was a little off when redrawing the components to position the whole node from the center bottom. I'll correct this to be the true center bottom.

Let me know what the difference is, so I can adjust Gas Properties.  Remember to do the same in SOM.

**Comment by pixelzoom** [comment]
> hoseCurvynessMultiplier: 1

Whoops, that's spelled wrong, should be `hoseCurvinessMultiplier`.  Or just `hoseCurviness` would probably be fine.

**Comment by pixelzoom** [comment]
You should probably verify that `hoseCurvinessMultiplier` behaves correctly when the hose attachment point is below the pump's origin.

**Comment by chrisklus** [comment]
>And you should probably verify that hoseCurvinessMultiplier behaves correctly when the hose attachment point is below the pump's origin.

yeah i'm betting it won't ha...

**Comment by pixelzoom** [comment]
Hmmm, why is this issue closed? Reopening.

**Comment by chrisklus** [comment]
Oops, probably me by mistake, thanks.

**Comment by pixelzoom** [comment]
Nope, it was me 6 days ago. Doh!

**Comment by chrisklus** [comment]
I've added `hoseCurviness` as an option, and to my surprise, it worked well when the bike pump was above the hose attachment point. If the bike pump is too close in the x-direction then the hose overlaps with the pump base, but I verified that that can be fixed by just increasing the `hoseCurviness` value.

<img height="284" src="https://user-images.githubusercontent.com/5618316/57051950-59fdf380-6c41-11e9-8276-138eae29c17e.png">  <img height="284" src="https://user-images.githubusercontent.com/5618316/57051970-6eda8700-6c41-11e9-94ab-464a2ef1a447.png">

I also updated the SOM usage to the previously used value and @pixelzoom's preference for Gas Properties.

**Comment by pixelzoom** [comment]
@chrisklus `hoseCurviness` looks great, thanks for doing that.   Do we need an assertion to verify that it's > 0?  

Another thing we need related to BicyclePumpNode is a demo on SCENERY_PHET/ComponentsScreenView.

**Comment by pixelzoom** [comment]
Self-unassigning. Feel free to assign back to me when there's more to review.

**Comment by pixelzoom** [comment]
Reducing priority to medium, since this is now usable in GP and SOM.  

Remaining work is items (4) and (5) in https://github.com/phetsims/states-of-matter/issues/217#issuecomment-488186493.   Item (5) is probably the more important to address sooner rather than later.

**Comment by pixelzoom** [comment]
@chrisklus we're having some problems with BicyclePumpNode over in https://github.com/phetsims/gas-properties/issues/62.  

From https://github.com/phetsims/gas-properties/issues/62#issuecomment-489337799, another thing we need to look at:
> The other odd thing that I've discovered about BicyclePumpNode is that it will only add particles one at a time. That's going to be generally problematic in this sim, because we want to add groups of particles simultaneous, with a gaussian distribution of initial temperature values, so that we don't end up with "waves" of particles.

The problematic code is:

```js
                // inject a particle
                if ( rangeProperty.value.max - numberProperty.value > 0 && enabledProperty.get() ) {
                  numberProperty.value++;
                }
```

For Gas Properties, we need something like `numberProperty.value += numberAddedOnThisDrag`.

Does SOM require adding particles one at a time?  
Do we need to add an option that determines how `numberProperty` is modified?

**Comment by chrisklus** [comment]
>Does SOM require adding particles one at a time?

Yes, I think it would be weird if we changed the behavior to match what you're requesting for Gas Properties. Perhaps @arouinfar would have more thoughts on this.

>Do we need to add an option that determines how numberProperty is modified?

Yeah, that sounds like the best way to go. 

>we need something like `numberProperty.value += numberAddedOnThisDrag`

There's some room for variation here - how do you want me to implement it? Here's some options I've thought about:

1. We could add the "acquired" particles at the `end` drag event. But if you pumped the handle up an down repeatedly, all of the particles would be added at once at the end, instead of at the end of each pumping motion.

2. We could add the acquired particles when the handle is in the same place for two `drag` events in a row. This probably wouldn't happen that naturally, and would likely result in a similar behavior to (1).

3. We could add the acquired particles when the handle reaches the "closed" end of its range. This could be nice, but a user probably shouldn't be required to drag the handle all the way down to add particles before starting another drag motion.

4. We could add the acquired particles whenever the handle changes direction to an upwards motion. This option makes the most sense to me, as the user is given the control to define how large of a drag they want per pumping motion, and particles are added immediately at the end of that pumping motion. The code for that could look something like this:

```js

if ( this.lastHandlePosition !== null ) {
  const travelDistance = handlePosition - this.lastHandlePosition;
  if ( travelDistance > 0 ) {

    // This motion is in the downward direction, so add its distance to the pumping distance.
    pumpingDistanceAccumulation += travelDistance;
    if( options.addParticlesOneAtATime ) {
      while ( pumpingDistanceAccumulation >= pumpingDistanceRequiredToAddParticle ) {

        // inject a particle
        if ( rangeProperty.value.max - numberProperty.value > 0 && enabledProperty.get() ) {
          numberProperty.value++;
        }
        pumpingDistanceAccumulation -= pumpingDistanceRequiredToAddParticle;
      }
    }
  }

  // This motion is in the upward direction
  else {
    if( !options.addParticlesOneAtATime ) {
      numberProperty.value += pumpingDistanceAccumulation / pumpingDistanceRequiredToAddParticle;
    }
    pumpingDistanceAccumulation = 0;
  }
}

```

**Comment by chrisklus** [comment]
@pixelzoom I tried out option (4) from https://github.com/phetsims/states-of-matter/issues/217#issuecomment-490162681, and the behavior is looking nice. I'm also no longer seeing any particles sticking at the hose attachment point (as described in https://github.com/phetsims/gas-properties/issues/62#issue-440271129).

I added nested options for the drag listener - let me know if you think that is unnecessary or too much exposure.

**Comment by pixelzoom** [comment]
(4) results in very odd behavior.  

Example 1:
1. Grab the pump handle.
2. Pull the pump handle up, then push the pump handle 90% of the way down.
3. Stop for 5 seconds, do not release the pump handle. Nothing comes out of the pump.
4. Pull the pump handle up.  Particles come out of the pump.

Example 2:
1. Grab the pump handle.
2. Pull the pump handle up, then push the pump handle 90% of the way down.
3. Stop for 5 seconds, do not release the pump handle. Nothing comes out of the pump.
4. Release the pump handle.  Nothing comes out of the pump.

So waiting until the handle is pulled up is not a good strategy.

**Comment by pixelzoom** [comment]
I think the next thing we should do is add BicyclePumpNode to scenery-phet demo, so that we're testing it in a test harness, vs in a sim that may have problems.

**Comment by chrisklus** [comment]
Yeah I'm confused why my edit to option (4) works for when the pump is all the way down but not stopped at an arbitrary place.

**Comment by chrisklus** [comment]
@pixelzoom I worked with @SaurabhTotey today and we finished (5) from https://github.com/phetsims/states-of-matter/issues/217#issuecomment-488186493 and updated the offset in your usage so that the pumps in Gas Properties are in the same place (the change was +11).


Remaining work for this issue:

- [ ] Wait to hear from @ariel-phet about (4) in https://github.com/phetsims/states-of-matter/issues/217#issuecomment-488186493.
- [ ] Add BicyclePumpNode to scenery-phet demo

**Comment by chrisklus** [comment]
>Yeah I'm confused why my edit to option (4) works for when the pump is all the way down but not stopped at an arbitrary place.

I understand now. When the pump handle is dragged all the way down, the drag listener is still firing because you typically drag your mouse past the bottom of the range, so the drag events know that the handle is not moving in a downward direction. But in @pixelzoom's examples in https://github.com/phetsims/states-of-matter/issues/217#issuecomment-490289661, the mouse is not moving, so there are no drag events after the most recent downward motion.

Any ideas on how to handle this @pixelzoom?

**Comment by arouinfar** [comment]
> >Does SOM require adding particles one at a time?

>Yes, I think it would be weird if we changed the behavior to match what you're requesting for Gas Properties. Perhaps @arouinfar would have more thoughts on this.

I'm not sure, actually. I know the model for SOM can't handle a sudden influx of particles, which is why the BicyclePumpNode only adds a max of 3 particles/pump. That said, I don't know if anything bad would happen if all 3 of those particles were added at the same time. From a pedagogical perspective, adding the particles simultaneously would be just fine. From a model perspective, I don't know. That'd be a question from @jbphet.

**Comment by pixelzoom** [comment]
@chrisklus in the above commit I changed how option `addParticlesOneAtATime` is handled.  I also moved the default for `dragListenerOptions` into `HandleNodeDragListener` constructor.  This seems to be working nicely in both SOM and GP.   Review and let me know if you see any issues.

I'm still seeing one particle get stuck where it enters the container at the hose.  See screenshot below.  No idea why this is happening and it's intermittent.  Any ideas?

![screenshot_1181](https://user-images.githubusercontent.com/3046552/57408675-81097780-71a3-11e9-9f65-6abcac317893.jpg)

**Comment by pixelzoom** [comment]
I said:
> This seems to be working nicely in both SOM and GP.

Guess I have to eat my words.  Temperature in the GP container is still behaving erratically and I'm not sure why.

**Comment by jbphet** [comment]
> Does SOM require adding particles one at a time?

The main reasons that the particles are not added all at once are:
+ Since the particles originate at the same place in the container, adding them at the same time also adds them at essentially the same **place**.  The repulsion between the particles is based on the distance between the particles, and when this distance approaches zero, the repulsion approaches infinity, and the model - not surprisingly - doesn't handle this particularly well.  The main consequence is that particles get going so fast that they end up outside the container.
+ It looks more real, at least to me, to have them come in sequentially.  I mean, it's not like bike pumps send bursts of air into a tire.  It's a stream of air.
+ The temperature gets recalculated when particles are added, and adding them all at once would likely cause more jumpy temperature readings.

Based on the above, I feel strongly that it should continue to work the way it has in the past, i.e. particles should be injected sequentially, not in bursts.

**Comment by pixelzoom** [comment]
Somes like all needs are met with option `addParticlesOneAtATime`.

**Comment by pixelzoom** [comment]
@chrisklus The stuck particle appears to have been a Gas Properties problem.  Details in https://github.com/phetsims/gas-properties/issues/62#issuecomment-491059437.  But in summary, it was possible to create a particle with zero initial speed, because the initial temperature was incorrect.

**Comment by ariel-phet** [comment]
In regards to https://github.com/phetsims/states-of-matter/issues/217#issuecomment-488186493 if there is an option for height, it should allow us to scale the height, not just the overall size of the pump. In future sims I could see such an option being useful. 

However, I am guessing it will be a nontrivial amount of rework this option. I will leave it to @chrisklus and @pixelzoom to decide if they want to add this option now, or (since I don't think either current sim needs such an option), perhaps make a separate issue for this work to be done at some point in the future, or punt until it is desired for a particular sim.

**Comment by chrisklus** [comment]
@pixelzoom after a bit of confusion and thought, your changes from https://github.com/phetsims/scenery-phet/commit/872860a38ac811394eeae94a8ea107dd4adb896e make sense and look good. I guess I didn't understand that literally the only issue you were trying to change was the uniform-looking wave of particles that appeared when BPN was quickly pumped. My understanding was that you wanted the particles to be batched per pumping stroke, not per drag event. Batching per drag event is a very subtle change from the original implementation. My misinterpretation of `numberAddedOnThisDrag` in https://github.com/phetsims/states-of-matter/issues/217#issuecomment-489338309 must have been how I got so far off track :)

Regarding https://github.com/phetsims/states-of-matter/issues/217#issuecomment-492445608 - I agree that it will likely need to be done one day, but I'm thinking it could be a while before another bike pump is needed in a sim *and* it needs to be scaled differently. So I think it's pretty low priority. 

My one reservation for pushing it off is that there is some scaling functionality already built into the code - it's just bad (the whole pump stretches instead of just the body tube changing size), it's only half working (the handle doesn't scale with the rest of the pump and the hose attachment becomes misaligned), and therefore it's not at all useful. I would feel better if there wasn't a way to stretch the width and height at all instead of the current state that it's in.

What do you think @pixelzoom?

**Comment by pixelzoom** [comment]
@chrisklus asked:
> ... What do you think @pixelzoom?

I agree that the current implementation is not a good way the way to handle sizing of the pump, and it's only a matter of time before it will need to be redone.  That said, I don't have time to work on this, and it's not needed for Gas Properties.  So I will defer the decision to you and @ariel-phet.

**Comment by chrisklus** [comment]
Okay, I'm just going to do it then. Can you assign a priority please @ariel-phet.

**Comment by pixelzoom** [comment]
BicyclePumpNode is failing in Gas Properties.  It's allowing me to pump in more particles than `rangeProperty`.   Still no demo in scenery-phet, so I'm going to add one.

**Comment by chrisklus** [comment]
Ah, looks like you skipped over the range check in https://github.com/phetsims/scenery-phet/commit/872860a38ac811394eeae94a8ea107dd4adb896e.

**Comment by pixelzoom** [comment]
Where are those checks skipped?  I'm not seeing it.  The relevant code is at line 635:

```js
if ( rangeProperty.value.max - numberProperty.value > 0 && enabledProperty.get() ) {
  if ( options.addParticlesOneAtATime ) {
    numberProperty.value++;
  }
  else {
    numberOfBatchParticles++;
  }
}
```

**Comment by pixelzoom** [comment]
Ah... The check is using `numberProperty.value`, which is not updated when `!options.addParticlesOneAtATime`.

**Comment by pixelzoom** [comment]
Range check fixed in the above commit.  Seems to be working fine in Gas Properties and scenery-phet demo.  I'm less familiar with SOM; I pumped until the pump was empty and didn't experience any problem.  @chrisklus could you check my work?

**Comment by pixelzoom** [comment]
@chrisklus you might also want to review `demoBicyclePumpNode` in scenery-phet demo.

**Comment by chrisklus** [comment]
>Range check fixed in the above commit. Seems to be working fine in Gas Properties and scenery-phet demo. I'm less familiar with SOM; I pumped until the pump was empty and didn't experience any problem. @chrisklus could you check my work?

Looks perfect. SOM will just work the same as before since `numberOfBatchParticles` is initialized to 0.

**Comment by chrisklus** [comment]
>@chrisklus you might also want to review `demoBicyclePumpNode` in scenery-phet demo.

Looks great, thanks for doing that.

**Comment by pixelzoom** [comment]
I believe work on this has been done for awhile, so I'm going to close this issue.  If that's not the case, @chrisklus please reopen.
